### PR TITLE
PT-162340809 Fix HTTP request validation result

### PR DESCRIPTION
**Pull request by jur0**
[PT-162340809](https://www.pivotaltracker.com/n/projects/2124891/stories/162340809)

The current default config of lager:
```
(epoch@localhost)8> lager:status().
Lager status:
File log/epoch_metrics.log (epoch_metrics_lager_event) at level info
File log/epoch_mining.log (epoch_mining_lager_event) at level info
File log/epoch_pow_cuckoo.log (epoch_pow_cuckoo_lager_event) at level debug
Console (epoch_sync_lager_event) at level info
File log/epoch_sync.log (epoch_sync_lager_event) at level debug
Console (lager_event) at level info
File log/epoch.log (lager_event) at level debug
```
It'd be good to set different log levels for the production system so debug, info and notice(?) messages are not logged by default.

### CODE REVIEW

File: apps/aehttp/src/aehttp_api_handler.erl [hunk 1]
```diff
@@ -63,8 +63,8 @@ handle_request_json(Req0, State = #state{
             Req = cowboy_req:reply(400, #{}, Body, Req1),
```
Comment (lucafavatella) [inline apps/aehttp/src/aehttp_api_handler.erl]: In https://github.com/aeternity/epoch/pull/1876/files I amended this warning ^^^ to info, idea being that identified error in request shall not warn the operator (but the user, that is notified via HTTP response body).

File: apps/aehttp/src/aehttp_api_handler.erl [hunk 2]
```diff
@@ -63,8 +63,8 @@ handle_request_json(Req0, State = #state{
...
             {stop, Req, State}
     catch error:Error ->
-            lager:error("Unexpected validate result: ~p / ~p",
```
Comment (lucafavatella) [inline apps/aehttp/src/aehttp_api_handler.erl]: It seems to me that this shall be kept as strictly-more-than-info log, on the basis that unidentified error shall be notified to the operator as maybe client request is valid but validator in server handler is incorrect as not catering for case.

**Comment by jur0** [comment]
The issue was that the request body wasn't read fully as described in: https://ninenines.eu/docs/en/cowboy/2.2/manual/cowboy_req.read_body/

File: apps/aehttp/src/aehttp_api_handler.erl [hunk 3]
```diff
@@ -63,8 +63,8 @@ handle_request_json(Req0, State = #state{
...
-                        [Error, erlang:get_stacktrace()]),
+            lager:warning("Unexpected validate result: ~p / ~p",
```
Comment (lucafavatella) [inline apps/aehttp/src/aehttp_api_handler.erl]: 👍

File: apps/aehttp/src/aehttp_api_validate.erl [hunk 4]
```diff
@@ -211,6 +211,11 @@ get_param_value("path", Name, Req) ->
     Value = cowboy_req:binding(to_binding(Name), Req),
     {ok, Value, Req}.
 
+read_body(Req0, Acc) ->
+    case cowboy_req:read_body(Req0) of
+        {ok, Data, Req} -> {ok, <<Acc/binary, Data/binary>>, Req};
+        {more, Data, Req} -> read_body(Req, <<Acc/binary, Data/binary>>)
```
Comment (lucafavatella) [inline apps/aehttp/src/aehttp_api_validate.erl]: Can a malicious client performs a 1GB request to the node?
Can a malicious client stream a neverending request to the node?

File: apps/aehttp/src/aehttp_api_validate.erl [hunk 5]
```diff
@@ -188,7 +188,12 @@ prepare_param_({"maximum", Max}, Value, Name, _) ->
     end.
 
 get_param_value("body", _, Req0) ->
-    case cowboy_req:read_body(Req0) of
+    %% Cowboy will attempt to read up to ~5MB of data for up to 10s. The call
+    %% will return when there is up to ~5MB of data or at the end of the 10s
+    %% period. If there is more data to read (after reading the initial 5MB),
+    %% {more, Body, Req} is returned and we can consider the request body to
+    %% be too big and the request is invalid.
+    case cowboy_req:read_body(Req0, #{length => 5000000, period => 10000}) of
```
Comment (lucafavatella) [inline apps/aehttp/src/aehttp_api_validate.erl]: I see we use cowboy 2.2.2
I see [defaults documented as 8MB and 15 seconds](https://ninenines.eu/docs/en/cowboy/2.2/manual/cowboy_req.read_body/). Re units, I see code [8000000 (B) and 15000 (ms)](https://github.com/ninenines/cowboy/blob/2.2.2/src/cowboy_req.erl#L462-L463).
Ok.

**Comment by lucafavatella** [comment]
> [PT-162340455](https://www.pivotaltracker.com/n/projects/2124891/stories/162340455)

You mean https://www.pivotaltracker.com/story/show/162340809 ?

**Comment by jur0** [comment]
> > [PT-162340455](https://www.pivotaltracker.com/n/projects/2124891/stories/162340455)
> 
> You mean https://www.pivotaltracker.com/story/show/162340809 ?

Yes. Updated the PR title, the branch name is ok.

**Comment by lucafavatella** [comment]
> /me drafting release notes proposal.

@jur0  https://github.com/aeternity/epoch/pull/1896

**Pull request merged**